int32 num
int32[3] size
int32[] index
float64[] val